### PR TITLE
Make allocators, PAPI, and sanitizers targets, support ASAN with custom allocators

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Make jemalloc & tcmalloc imported targets
- Get the linking correct for custom allocators (they need to be linked before anything else)
- Make PAPI an imported target
- Make sanitizers an imported target. Still need to control link flags with a variable because CMake doesn't support targets having `INTERFACE_LINK_OPTIONS` until v3.13
- Allow using custom allocators with ASAN. Getting the linking order correct seems to have resolved any issues with this.

Depends on:
- [x] #2149 because I didn't bother getting the allocators right the compilation tests

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [x] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: cmake/AddSpectreExecutable.cmake [hunk 1]
```diff
@@ -5,17 +5,34 @@
...
   set_target_properties(
     ${TARGET_NAME}
     PROPERTIES
     RULE_LAUNCH_LINK "${CMAKE_BINARY_DIR}/tmp/WrapExecutableLinker.sh"
     LINK_DEPENDS "${CMAKE_BINARY_DIR}/tmp/WrapExecutableLinker.sh"
     )
-  target_link_libraries(
-    ${TARGET_NAME}
-    PRIVATE
-    SpectreFlags
```
Comment (nilsvu) [inline cmake/AddSpectreExecutable.cmake]: Why do you remove linking with `SpectreFlags`?

File: cmake/SetupSanitizers.cmake [hunk 2]
```diff
@@ -2,37 +2,65 @@
...
 option(ASAN "Add AddressSanitizer compile flags" OFF)
+
+# We handle the sanitizers using targets for the compile options, but modify
+# CMAKE_EXE_LINKER_FLAGS for the linker flags because the sanitizers should
+# only be linked into the final executable and CMake doesn't support linker
+# interface flags before CMake 3.13.
+add_library(Sanitizers INTERFACE)
+add_library(AddressSanitizer INTERFACE)
+add_library(UbIntegerSanitizer INTERFACE)
+add_library(UbUndefinedSanitizer INTERFACE)
```
Comment (nilsvu) [inline cmake/SetupSanitizers.cmake]: Namespace them? E.g. `Sanitizers::Address`

**Comment by nilsdeppe** [comment]
Thanks for the review @nilsleiffischer! Pushed fixups :)

**Review by nilsvu — COMMENTED** [review]
please squash-ahead!

**Comment by nilsdeppe** [comment]
okay, squashed

**Review by kidder — DISMISSED** [review]

**Review by nilsvu — DISMISSED** [review]
Do you need to rebase on develop to drop the first commit?

**Comment by nilsdeppe** [comment]
No, since that PR was never changed the commit will just disappear

**Comment by nilsdeppe** [comment]
Rebased on develop. Thanks for the reviews :)

**Review by nilsvu — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@wthrowe can you check this on your setup

**Comment by wthrowe** [comment]
Works for me.

**Pull request merged**